osf/metrics/reporters/institutional_users.py

```diff
@@ -0,0 +1,135 @@
+    yearmonth: YearMonth
+    before_datetime: datetime.datetime
+    report: InstitutionalUserReport = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        _affiliation = self.user.get_institution_affiliation(self.institution._id)
+        self.report = InstitutionalUserReport(
+            report_yearmonth=self.yearmonth,
+            institution_id=self.institution._id,
+            user_id=self.user._id,
+            user_name=self.user.fullname,
+            department_name=(_affiliation.sso_department or None),
+            month_last_login=YearMonth.from_date(self.user.date_last_login),
+            account_creation_date=YearMonth.from_date(self.user.created),
+            orcid_id=self.user.get_verified_external_id('ORCID', verified_only=True),
+            public_project_count=self._public_project_queryset().count(),
+            private_project_count=self._private_project_queryset().count(),
+            public_registration_count=self._public_registration_queryset().count(),
+            embargoed_registration_count=self._embargoed_registration_queryset().count(),
+            public_file_count=self._public_osfstorage_file_queryset().count(),
+            published_preprint_count=self._published_preprint_queryset().count(),
+            storage_byte_count=self._storage_byte_count(),
+        )
+
+    def _node_queryset(self):
+        _institution_node_qs = self.institution.nodes.filter(
+            created__lt=self.before_datetime,
+            is_deleted=False,
+        ).exclude(spam_status=SpamStatus.SPAM)
+        return osfdb.Node.objects.get_nodes_for_user(
+            user=self.user,
+            base_queryset=_institution_node_qs,
+        )
+
+    def _public_project_queryset(self):
+        return self._node_queryset().filter(
+            type='osf.node',  # `type` field from TypedModel
+            is_public=True,
+            root_id=F('pk'),  # only root nodes
+        )
+
+    def _private_project_queryset(self):
+        return self._node_queryset().filter(
+            type='osf.node',  # `type` field from TypedModel
+            is_public=False,
+            root_id=F('pk'),  # only root nodes
+        )
+
+    def _public_registration_queryset(self):
+        return self._node_queryset().filter(
+            type='osf.registration',  # `type` field from TypedModel
+            is_public=True,
+            root_id=F('pk'),  # only root nodes
+        )
+
+    def _embargoed_registration_queryset(self):
+        return self._node_queryset().filter(
+            type='osf.registration',  # `type` field from TypedModel
+            is_public=False,
+            root_id=F('pk'),  # only root nodes
+            embargo__end_date__gte=self.before_datetime,
+        )
+
+    def _published_preprint_queryset(self):
+        if not hasattr(osfdb.Preprint, 'affiliated_institutions'):
+            return osfdb.Preprint.objects.none()  # HACK: preprints affiliation project still in-progress
```

Comment:
nit: if I was going to be a real stickler I would say this should match the creator affiliation numbers to match current SHARE, but that's really minor and probably won't confuse anyone. NBD